lib/translator/write.cc

```diff
@@ -0,0 +1,217 @@
+      prchk = 0b111;
+      break;
+    case 0b010:
+      pract = 0;
+      prchk = 0b011;
+      break;
+    case 0b011:
+      pract = 0;
+      prchk = 0b00;
+      break;
+    case 0b100:
+      pract = 0;
+      prchk = 0b100;
+      break;
+    default:
+      // All other codes shall result in command termination with CHECK
+      // CONDITION status, ILLEGAL REQUEST sense key, and ILLEGAL FIELD IN CDB
+      // additional sense code.
+      DebugLog("Invalid WriteProtect Code for PRInfo.");
+      return StatusCode::kFailure;
+  }
+
+  pr_info = prchk | (pract << 3);
+  return StatusCode::kSuccess;
+}
+
+// Converts transfer_length in units of logical blocks to units of pages
+uint32_t GetTransferLengthPages(uint16_t transfer_length, uint32_t page_size,
+                                uint32_t lba_size) {
+  uint64_t transfer_length_bytes = transfer_length * lba_size;
+  return transfer_length_bytes / page_size +
+         ((transfer_length_bytes % page_size == 0) ? 0 : 1);
+}
+
+// This function populates the nvme::GenericQueueEntryCmd object with fields
+// that are common to all Write Commands (6, 10, 12, 16) Refer to Section 5.7
+// (https://nvmexpress.org/wp-content/uploads/NVM_Express_-_SCSI_Translation_Reference-1_5_20150624_Gold.pdf)
+StatusCode LegacyWrite(NvmeCmdWrapper& nvme_wrapper, Allocation& allocation,
+                       uint32_t nsid, uint32_t transfer_length_pages) {
+  StatusCode status_code = allocation.SetPages(transfer_length_pages, 0);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  nvme_wrapper.cmd = {.opc = static_cast<uint8_t>(nvme::NvmOpcode::kWrite),
+                      .psdt = 0,  // prps are used
+                      .nsid = nsid};
+
+  nvme_wrapper.cmd.dptr.prp.prp1 = allocation.data_addr;
+
+  return status_code;
+}
+
+// Builds NVMe cdw 12 for Write10, Write12, Write16 translations
+uint32_t BuildCdw12(uint16_t transfer_length, uint8_t prinfo, bool fua) {
+  // cdw12 nlb bits 15:00 (zero based field), prinfo bits 29:26, fua bit 30
+  return (static_cast<uint32_t>(fua) << 30) |
+         (static_cast<uint32_t>(prinfo) << 26) | (transfer_length - 1);
+}
+
+StatusCode Write(bool fua, uint8_t wrprotect, uint32_t transfer_length,
+                 NvmeCmdWrapper& nvme_wrapper, Allocation& allocation,
+                 uint32_t nsid, uint32_t page_size, uint32_t lba_size) {
+  if (transfer_length == 0) {
+    DebugLog("NVMe read command does not support transfering zero blocks\n");
+    return StatusCode::kNoTranslation;
+  }
+
+  transfer_length &= 0xffff;  // truncate to 16 bits
+  StatusCode status_code =
+      LegacyWrite(nvme_wrapper, allocation, nsid,
+                  GetTransferLengthPages(transfer_length, page_size, lba_size));
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  uint8_t pr_info;
+  status_code = BuildPRInfo(wrprotect, pr_info);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
```

Comment:
dbglg

Reply:
the function from which this error is coming from already prints a debug message. So, it would cause duplication